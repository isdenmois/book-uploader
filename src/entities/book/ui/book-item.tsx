import React, { FC, memo } from 'react'
import { StyleSheet } from 'react-native'
import FastImage from 'react-native-fast-image'

import { BookItem as IBookItem } from 'shared/api'
import { getImageUrl } from 'shared/api/book-image'
import { Box, FileIcon, Item, Text } from 'shared/ui'

interface Props {
  item: IBookItem
  onPress: (item: IBookItem) => void
}

export const BookItem: FC<Props> = memo<Props>(
  ({ item, onPress }) => {
    const other = [item.size, item.lang, item.translation].filter(p => p).join(', ')
    const ext = item.ext.replace('.zip', '')
    const imageUrl = getImageUrl(item.type, item.imageUrl)

    return (
      <Item mb={3} onPress={() => onPress(item)} testID='book-item'>
        {!!imageUrl && (
          <Box borderRadius={4} width={40} height={50} overflow='hidden' backgroundColor='searchSelected'>
            <FastImage style={s.image} source={{ uri: imageUrl }} resizeMode='cover' />

            <Text fontSize={12} fontWeight='bold' textAlign='center' color='white' style={s.text}>
              {ext}
            </Text>
          </Box>
        )}
        {!imageUrl && <FileIcon color='searchSelected' text={ext} />}

        <Item.Text suptitle={item.authors} title={item.title} subtitle={other} />
      </Item>
    )
  },
  (prev, next) => prev.item === next.item,
)

const s = StyleSheet.create({
  image: {
    width: 40,
    height: 50,
  },
  text: {
    backgroundColor: '#0005',
    bottom: 0,
    left: 0,
    right: 0,
    position: 'absolute',
  },
})
