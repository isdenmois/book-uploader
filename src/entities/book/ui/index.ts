export * from './book-item'
export * from './book-list'
