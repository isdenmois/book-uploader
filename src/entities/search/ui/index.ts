export * from './search-filters'
export * from './book-list'
