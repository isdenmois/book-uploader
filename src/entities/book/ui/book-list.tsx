import React, { FC } from 'react'
import { FlatList, Image, StyleSheet, View } from 'react-native'

import { BookItem as IBookItem } from 'shared/api'
import { ActivityIndicator, Box, Text } from 'shared/ui'
import { Images } from 'shared/assets'

import { BookItem } from './book-item'

type Props = {
  books: IBookItem[]
  isLoading: boolean
  onDownload: (item: IBookItem) => void
}

export const BookList: FC<Props> = ({ onDownload, books, isLoading }) => {
  if (isLoading) {
    return (
      <Box flex={1} justifyContent='center'>
        <ActivityIndicator color='searchSelected' size='large' />
      </Box>
    )
  }

  if (!books)
    return (
      <View style={s.empty}>
        <Image style={s.image} source={Images.startSearch} fadeDuration={0} resizeMode='contain' />
        <Text style={s.startText} color='search'>
          Start search books to see a result
        </Text>
      </View>
    )

  return (
    <FlatList
      data={books}
      keyExtractor={item => item.link}
      renderItem={({ item }) => <BookItem item={item} onPress={onDownload} />}
      style={s.list}
      contentContainerStyle={s.listContent}
      ListEmptyComponent={
        <View style={s.empty}>
          <Image style={s.image} source={Images.noResult} fadeDuration={0} resizeMode='contain' />
          <Text style={s.emptyText} color='search'>
            Nothing has found
          </Text>
        </View>
      }
    />
  )
}

const s = StyleSheet.create({
  empty: {
    flex: 1,
    marginTop: 30,
  },
  image: {
    width: '100%',
  },
  startText: {
    fontSize: 20,
    textAlign: 'center',
  },
  emptyText: {
    fontSize: 24,
    textAlign: 'center',
    marginTop: 20,
  },
  list: {
    marginTop: 15,
  },
  listContent: {
    paddingHorizontal: 15,
  },
})
