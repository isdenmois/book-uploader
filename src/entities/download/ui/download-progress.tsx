import React from 'react'
import { useStore } from 'effector-react'

import { ProgressBar, Text } from 'shared/ui'
import { $progress } from '../model'

export const DownloadProgress = () => {
  const progress = useStore($progress)
  if (progress < 0) {
    return null
  }

  if (progress === 0) {
    return <Text variant='secondary'>Connection...</Text>
  }

  return <ProgressBar progress={progress} color='searchSelected' />
}
