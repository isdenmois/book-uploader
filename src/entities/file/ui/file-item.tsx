import React, { FC } from 'react'
import { TouchableOpacity } from 'react-native'

import { FileIcon, Item, ProgressBar, TimesIcon } from 'shared/ui'
import { useConfirm } from 'shared/utils'
import type { FileData } from '../model'

interface Props {
  file: FileData
  disabled?: boolean
  onRemove: (file: FileData) => void
  onParse: (file: FileData) => void
  onShare: (file: FileData) => void
}

export const FileItem: FC<Props> = ({ file, disabled, onRemove, onParse, onShare }) => {
  const parseDisabled = disabled || file.isParsed
  const removeFile = useConfirm('Remove file?', file.title || file.id, () => onRemove(file))

  return (
    <Item
      mb={3}
      onPress={() => onShare(file)}
      onLongPress={parseDisabled ? null : () => onParse(file)}
      disabled={disabled}
      progress={!file.error && <ProgressBar progress={file.progress} color='uploadSelected' />}
      testID={'file-item-' + file.id}
    >
      <FileIcon size={34} color='uploadSelected' text={file.ext} />

      <Item.Text suptitle={file.author} title={file.title} error={file.error} />

      {!disabled && (
        <TouchableOpacity onPress={removeFile} testID='removeFile'>
          <TimesIcon />
        </TouchableOpacity>
      )}
    </Item>
  )
}
