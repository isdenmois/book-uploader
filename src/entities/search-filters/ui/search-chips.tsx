import React, { FC } from 'react'
import { ScrollView, StyleSheet, View } from 'react-native'
import { useStore } from 'effector-react'

import { Chip } from 'shared/ui'

import { $searchFilters, setEpub, setFb2, setFlibusta, setZLib } from '../model'

interface Props {
  disabled?: boolean
}

export const SearchChips: FC<Props> = ({ disabled }) => {
  const { extension, type } = useStore($searchFilters)
  const isZLib = type === 'ZLIB'

  return (
    <ScrollView horizontal keyboardShouldPersistTaps='always'>
      <Chip title='Z-Library' selected={type === 'ZLIB'} onPress={setZLib} disabled={disabled} />
      <Chip title='Flibusta' selected={type === 'FLIBUSTA'} onPress={setFlibusta} disabled={disabled} />

      <View style={s.placeholder} />

      {isZLib && <Chip title='EPUB' selected={extension === 'epub'} onPress={setEpub} disabled={disabled} />}
      {isZLib && <Chip title='FB2' selected={extension === 'fb2'} onPress={setFb2} disabled={disabled} />}
    </ScrollView>
  )
}

const s = StyleSheet.create({
  placeholder: { width: 20 },
})
