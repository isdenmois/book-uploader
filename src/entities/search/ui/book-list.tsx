import React, { FC } from 'react'
import { FlatList, Image, StyleSheet, View } from 'react-native'
import { useStore } from 'effector-react'

import { BookItem } from 'shared/api'
import { ActivityIndicator, Box, Text } from 'shared/ui'
import { Images } from 'shared/assets'

import { SearchItem } from './search-item'
import { $booksFound, fetchBookItemsFx } from '../model'

type Props = {
  onDownload: (item: BookItem) => void
}

export const BookList: FC<Props> = ({ onDownload }) => {
  const books = useStore($booksFound)
  const loading = useStore(fetchBookItemsFx.pending)

  if (loading) {
    return (
      <Box flex={1} justifyContent='center'>
        <ActivityIndicator color='searchSelected' size='large' />
      </Box>
    )
  }

  if (!books)
    return (
      <View style={s.empty}>
        <Image style={s.image} source={Images.startSearch} fadeDuration={0} resizeMode='contain' />
        <Text style={s.startText} color='search'>
          Start search books to see a result
        </Text>
      </View>
    )

  return (
    <FlatList
      data={books}
      keyExtractor={item => item.link}
      renderItem={({ item }) => <SearchItem item={item} onPress={onDownload} />}
      style={s.list}
      contentContainerStyle={s.listContent}
      ListEmptyComponent={
        <View style={s.empty}>
          <Image style={s.image} source={Images.noResult} fadeDuration={0} resizeMode='contain' />
          <Text style={s.emptyText} color='search'>
            Nothing has found
          </Text>
        </View>
      }
    />
  )
}

const s = StyleSheet.create({
  empty: {
    flex: 1,
    marginTop: 30,
  },
  image: {
    width: '100%',
  },
  startText: {
    fontSize: 20,
    textAlign: 'center',
  },
  emptyText: {
    fontSize: 24,
    textAlign: 'center',
    marginTop: 20,
  },
  list: {
    marginTop: 15,
  },
  listContent: {
    paddingHorizontal: 15,
  },
})
