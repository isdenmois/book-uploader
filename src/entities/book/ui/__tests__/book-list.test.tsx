import React from 'react'
import { act, fireEvent, render } from '@testing-library/react-native'

import { BookItem } from 'shared/api'
import { ThemeProvider } from 'shared/ui'

import { BookList } from '../book-list'

test('BookItem', async () => {
  const books: BookItem[] = [
    { ext: 'epub', type: 'ZLIB', title: 'First result', link: 'http://example.com' },
    { ext: 'epub', type: 'ZLIB', title: 'Second', link: 'http://ex.com' },
    { ext: 'epub', type: 'ZLIB', title: 'Third res', link: 'http://q.com' },
  ]
  const onDownload = jest.fn()

  const { rerender, toJSON, findAllByTestId, unmount } = render(
    <BookList books={null} isLoading onDownload={onDownload} />,
    {
      wrapper: ThemeProvider,
    },
  )

  expect(toJSON()).toMatchSnapshot()

  rerender(<BookList books={null} isLoading={false} onDownload={onDownload} />)
  expect(toJSON()).toMatchSnapshot()

  rerender(<BookList books={books} isLoading={false} onDownload={onDownload} />)
  expect(toJSON()).toMatchSnapshot()

  fireEvent.press((await findAllByTestId('book-item'))[0])

  expect(onDownload).toHaveBeenCalledWith(books[0])

  await act(async () => {
    unmount()
  })
})
