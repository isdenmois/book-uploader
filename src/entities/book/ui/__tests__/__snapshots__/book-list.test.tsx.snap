// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <ActivityIndicator
    color="#5396EE"
    size="large"
  />
</View>
`;

exports[`BookItem 2`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginTop": 30,
    }
  }
>
  <Image
    fadeDuration={0}
    resizeMode="contain"
    source={
      Object {
        "testUri": "../../../src/shared/assets/images/start-search.png",
      }
    }
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#8C93A3",
          "display": "flex",
          "fontSize": 16,
        },
        Object {
          "fontSize": 20,
          "textAlign": "center",
        },
      ]
    }
  >
    Start search books to see a result
  </Text>
</View>
`;

exports[`BookItem 3`] = `
<RCTScrollView
  ListEmptyComponent={
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 30,
        }
      }
    >
      <Image
        fadeDuration={0}
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/shared/assets/images/no-result.png",
          }
        }
        style={
          Object {
            "width": "100%",
          }
        }
      />
      <ForwardRef
        color="search"
        style={
          Object {
            "fontSize": 24,
            "marginTop": 20,
            "textAlign": "center",
          }
        }
      >
        Nothing has found
      </ForwardRef>
    </View>
  }
  contentContainerStyle={
    Object {
      "paddingHorizontal": 15,
    }
  }
  data={
    Array [
      Object {
        "ext": "epub",
        "link": "http://example.com",
        "title": "First result",
        "type": "ZLIB",
      },
      Object {
        "ext": "epub",
        "link": "http://ex.com",
        "title": "Second",
        "type": "ZLIB",
      },
      Object {
        "ext": "epub",
        "link": "http://q.com",
        "title": "Third res",
        "type": "ZLIB",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginTop": 15,
    }
  }
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "marginBottom": 24,
            "opacity": 1,
          }
        }
        testID="book-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
              },
            ]
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={50}
            bbWidth={40}
            focusable={false}
            height={50}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 50,
                  "width": 40,
                },
              ]
            }
            vbHeight={50}
            vbWidth={40}
            width={40}
          >
            <RNSVGGroup>
              <RNSVGGroup>
                <RNSVGPath
                  d="M24 0H2.4A2.394 2.394 90 000 2.388V48.8a2.394 2.394 90 002.388 2.4H36a2.394 2.394 90 002.4-2.388V14.4H25.6a1.6 1.6 90 01-1.6-1.6zm14.4 12.19v.61H27.2a1.6 1.6 90 01-1.6-1.6V0h.61a2.4 2.4 90 011.7.7l9.79 9.8a2.39 2.39 90 01.7 1.69z"
                  fill={4283668206}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
              <RNSVGText
                content={null}
                dx={Array []}
                dy={Array []}
                fill={4294967295}
                font={
                  Object {
                    "fontSize": "12",
                    "fontWeight": "bold",
                    "textAnchor": "middle",
                  }
                }
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                x={
                  Array [
                    "19",
                  ]
                }
                y={
                  Array [
                    "44",
                  ]
                }
              >
                <RNSVGTSpan
                  content="EPUB"
                  dx={Array []}
                  dy={Array []}
                  font={Object {}}
                  rotate={Array []}
                  x={Array []}
                  y={Array []}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                  "marginLeft": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#092D5D",
                    "display": "flex",
                    "fontSize": 16,
                  },
                ]
              }
            >
              First result
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "marginBottom": 24,
            "opacity": 1,
          }
        }
        testID="book-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
              },
            ]
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={50}
            bbWidth={40}
            focusable={false}
            height={50}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 50,
                  "width": 40,
                },
              ]
            }
            vbHeight={50}
            vbWidth={40}
            width={40}
          >
            <RNSVGGroup>
              <RNSVGGroup>
                <RNSVGPath
                  d="M24 0H2.4A2.394 2.394 90 000 2.388V48.8a2.394 2.394 90 002.388 2.4H36a2.394 2.394 90 002.4-2.388V14.4H25.6a1.6 1.6 90 01-1.6-1.6zm14.4 12.19v.61H27.2a1.6 1.6 90 01-1.6-1.6V0h.61a2.4 2.4 90 011.7.7l9.79 9.8a2.39 2.39 90 01.7 1.69z"
                  fill={4283668206}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
              <RNSVGText
                content={null}
                dx={Array []}
                dy={Array []}
                fill={4294967295}
                font={
                  Object {
                    "fontSize": "12",
                    "fontWeight": "bold",
                    "textAnchor": "middle",
                  }
                }
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                x={
                  Array [
                    "19",
                  ]
                }
                y={
                  Array [
                    "44",
                  ]
                }
              >
                <RNSVGTSpan
                  content="EPUB"
                  dx={Array []}
                  dy={Array []}
                  font={Object {}}
                  rotate={Array []}
                  x={Array []}
                  y={Array []}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                  "marginLeft": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#092D5D",
                    "display": "flex",
                    "fontSize": 16,
                  },
                ]
              }
            >
              Second
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "display": "flex",
            "marginBottom": 24,
            "opacity": 1,
          }
        }
        testID="book-item"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
              },
            ]
          }
        >
          <RNSVGSvgView
            align="xMidYMid"
            bbHeight={50}
            bbWidth={40}
            focusable={false}
            height={50}
            meetOrSlice={0}
            minX={0}
            minY={0}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                Object {
                  "flex": 0,
                  "height": 50,
                  "width": 40,
                },
              ]
            }
            vbHeight={50}
            vbWidth={40}
            width={40}
          >
            <RNSVGGroup>
              <RNSVGGroup>
                <RNSVGPath
                  d="M24 0H2.4A2.394 2.394 90 000 2.388V48.8a2.394 2.394 90 002.388 2.4H36a2.394 2.394 90 002.4-2.388V14.4H25.6a1.6 1.6 90 01-1.6-1.6zm14.4 12.19v.61H27.2a1.6 1.6 90 01-1.6-1.6V0h.61a2.4 2.4 90 011.7.7l9.79 9.8a2.39 2.39 90 01.7 1.69z"
                  fill={4283668206}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                />
              </RNSVGGroup>
              <RNSVGText
                content={null}
                dx={Array []}
                dy={Array []}
                fill={4294967295}
                font={
                  Object {
                    "fontSize": "12",
                    "fontWeight": "bold",
                    "textAnchor": "middle",
                  }
                }
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotate={Array []}
                x={
                  Array [
                    "19",
                  ]
                }
                y={
                  Array [
                    "44",
                  ]
                }
              >
                <RNSVGTSpan
                  content="EPUB"
                  dx={Array []}
                  dy={Array []}
                  font={Object {}}
                  rotate={Array []}
                  x={Array []}
                  y={Array []}
                />
              </RNSVGText>
            </RNSVGGroup>
          </RNSVGSvgView>
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                  "marginLeft": 8,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#092D5D",
                    "display": "flex",
                    "fontSize": 16,
                  },
                ]
              }
            >
              Third res
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
